From a4e0faf988489a4d0e4372896b5a7d527b18c17a Mon Sep 17 00:00:00 2001
From: gbhakteX <gurudattax.bhakte@intel.com>
Date: Fri, 18 Jan 2013 14:31:37 +0530
Subject: [PORT FROM MAIN] Audio-MW: [R4 All] Klocwork Fix required in Stagefright engine (2 Errors).

BZ: 81354

Klocwork issue #2817: Resource acquired to 'new(sizeof(FileSource))->mBuffer.mLock.mMutex'
at line 147 may be lost here. Also there are 3 similar errors on line(s) 193, 196.
Fix: 'source' is cleared before returning NULL, to avoid resource leak.
[DataSource.cpp | CreateFromURI() | line 193].

Klocwork issue #5255: Resource acquired to 'new(sizeof(FileSource))->mBuffer.mLock.mMutex'
at line 222 may be lost here. Also there is one similar error on line 310.
Fix: 'source' (only if it is FileSource) is cleared before returning from function
fetchFile(), to avoid resource leak.
[LiveSession.cpp | fetchFile() | line 298, 313].

Change-Id: I73f625a096658d7aa85294335bc8e694198cfbbc
Orig-Change-Id: I8afb6582af275e9577ba2882fc28899af9f5d7c1
Signed-off-by: gbhakteX <gurudattax.bhakte@intel.com>
Reviewed-on: http://android.intel.com:8080/87494
Reviewed-by: Kandasamy, Muthukumar <muthukumar.kandasamy@intel.com>
Reviewed-by: Anand, Jerome <jerome.anand@intel.com>
Reviewed-by: Krishna, Samaga <samaga.krishna@intel.com>
Reviewed-by: P C, SreekanthX <sreekanthx.p.c@intel.com>
Tested-by: Gupta, ArvindX K <arvindx.k.gupta@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
---
 media/libstagefright/DataSource.cpp           |    3 +++
 media/libstagefright/httplive/LiveSession.cpp |    6 ++++++
 2 files changed, 9 insertions(+), 0 deletions(-)

diff --git a/media/libstagefright/DataSource.cpp b/media/libstagefright/DataSource.cpp
index 507475e..95a6b05 100644
--- a/media/libstagefright/DataSource.cpp
+++ b/media/libstagefright/DataSource.cpp
@@ -227,6 +227,9 @@ sp<DataSource> DataSource::CreateFromURI(
     }
 
     if (source == NULL || source->initCheck() != OK) {
+        if (source != NULL) {
+            source.clear();
+        }
         return NULL;
     }
 
diff --git a/media/libstagefright/httplive/LiveSession.cpp b/media/libstagefright/httplive/LiveSession.cpp
index 817857e..23e0e65 100644
--- a/media/libstagefright/httplive/LiveSession.cpp
+++ b/media/libstagefright/httplive/LiveSession.cpp
@@ -300,6 +300,9 @@ status_t LiveSession::fetchFile(
                 maxBytesToRead);
 
         if (n < 0) {
+            if ((!strncasecmp(url, "file://", 7)) && (source != NULL)) {
+                source.clear();
+            }
             return n;
         }
 
@@ -312,6 +315,9 @@ status_t LiveSession::fetchFile(
 
     *out = buffer;
 
+    if ((!strncasecmp(url, "file://", 7)) && (source != NULL)) {
+        source.clear();
+    }
     return OK;
 }
 
-- 
1.7.4.1

